src/objective-c/GRPCClient/private/GRPCChannelPool.m

```diff
@@ -0,0 +1,387 @@
+  result ^= (unsigned int)(_callOptions.connectMaxBackoff * 1000);
+  result ^= _callOptions.additionalChannelArgs.hash;
+  result ^= _callOptions.PEMRootCertificates.hash;
+  result ^= _callOptions.PEMPrivateKey.hash;
+  result ^= _callOptions.PEMCertChain.hash;
+  result ^= _callOptions.hostNameOverride.hash;
+  result ^= _callOptions.transportType;
+  result ^= [_callOptions.logContext hash];
+  result ^= _callOptions.channelPoolDomain.hash;
+  result ^= _callOptions.channelID;
+
+  return result;
+}
+
+@end
+
+/**
+ * Time the channel destroy when the channel's calls are unreffed. If there's new call, reset the
+ * timer.
+ */
+@interface GRPCChannelCallRef : NSObject
+
+- (instancetype)initWithChannelConfiguration:(GRPCChannelConfiguration *)configuration
+                                destroyDelay:(NSTimeInterval)destroyDelay
+                               dispatchQueue:(dispatch_queue_t)dispatchQueue
+                              destroyChannel:(void (^)())destroyChannel;
+
+/** Add call ref count to the channel and maybe reset the timer. */
+- (void)refChannel;
+
+/** Reduce call ref count to the channel and maybe set the timer. */
+- (void)unrefChannel;
+
+@end
+
+@implementation GRPCChannelCallRef {
+  GRPCChannelConfiguration *_configuration;
+  NSTimeInterval _destroyDelay;
+  // We use dispatch queue for this purpose since timer invalidation must happen on the same
+  // thread which issued the timer.
+  dispatch_queue_t _dispatchQueue;
+  void (^_destroyChannel)();
+
+  NSUInteger _refCount;
+  NSTimer *_timer;
+}
+
+- (instancetype)initWithChannelConfiguration:(GRPCChannelConfiguration *)configuration
+                                destroyDelay:(NSTimeInterval)destroyDelay
+                               dispatchQueue:(dispatch_queue_t)dispatchQueue
+                              destroyChannel:(void (^)())destroyChannel {
+  if ((self = [super init])) {
+    _configuration = configuration;
+    _destroyDelay = destroyDelay;
+    _dispatchQueue = dispatchQueue;
+    _destroyChannel = destroyChannel;
+
+    _refCount = 0;
+    _timer = nil;
+  }
+  return self;
+}
+
+// This function is protected by channel pool dispatch queue.
+- (void)refChannel {
+  _refCount++;
+  if (_timer) {
+    [_timer invalidate];
+  }
+  _timer = nil;
+}
+
+// This function is protected by channel spool dispatch queue.
+- (void)unrefChannel {
+  self->_refCount--;
+  if (self->_refCount == 0) {
+    if (self->_timer) {
+      [self->_timer invalidate];
+    }
+    self->_timer = [NSTimer scheduledTimerWithTimeInterval:self->_destroyDelay
+                                                    target:self
+                                                  selector:@selector(timerFire:)
+                                                  userInfo:nil
+                                                   repeats:NO];
+  }
+}
+
+- (void)timerFire:(NSTimer *)timer {
+  dispatch_sync(_dispatchQueue, ^{
+    if (self->_timer == nil || self->_timer != timer) {
```

Comment:
Because it is possible that during the 30s delay, another call is created on the channel and then destroyed. In that case refcount is still 0 but we should not immediately destroy the channel.

Reply:
Ok, well in the move to dispatch_after you're going to loose the timer ivar to use as a flag. you might be better off just tracking the time since the last ref was made.